### ISSUE TITLE
Scaling apartment

### ISSUE THREAD

**Issue by vfonic** [influitive/apartment#420]
Hi,

First of all, thanks for incredible gem! :) (trully)

I'm building (yet another) SaaS CMS solution and would like to use a single rails app deployed on AWS, but use apartment to have separate database for each user. Each db schema will be the same, but I don't know how big certain tenant databases will grow.

For example, some users might just sign up and never login again. I'd have to create new tenant for such user. On the other hand, some tenants might become too big to "stay" in shared database.

What would you suggest as a good way of dynamically scaling databases? If I have each tenant on a different server, it could be too expensive for a lot of tenants that would just log in. On the other hand, if I share a single db with different schema (name) per tenant, how can I move the tenant to a different server once it grows too big?
Would it be possible to maybe move the data manually and then, on the next deployment, enable `with_multi_server_setup`, provide the same db server for "small" tenants and specify new, specific db server for "big" tenants?

Thanks!

Ref: https://github.com/influitive/apartment#tenants-on-different-servers

**Comment by Shaneprrlt** [comment]
I am interested in this issue too, but for different reasons. I am considering using Apartment as a solution for multi-tenancy in my app, but I started considering the cost of running migrations and cost of having potentially thousands of schemas once you start to scale. I checked into Heroku's docs to see if there were any limitations on number of schemas and while they do not limit it, they advise against using multiple-schema approach to multi-tenancy since their automated backup tool has performance issues around only 50 schemas.

https://devcenter.heroku.com/articles/heroku-postgresql#multiple-schemas

**Comment by mikecmpbll** [comment]
50 schemas? that seems ridiculously low. many people run production apps with many thousands of schemas afaik.

about scaling, we currently have a single production instance of MySQL (MariaDB) with about 4500 tenants. because it's mysql, it's 1 database per tenant. i'm not sure exactly but it's probably about 200GB total size.

we aim to shard at about 5000, and by shard i mean set up another DB server for the next 5000 or so. we'll be using the multi-server code in Apartment, hopefully, so that should prove a decent real world test of that (as the implementation is still quite experimental.)

**Comment by ozgune** [comment]
Hi @vfonic, @Shaneprrlt, @mikecmpbll,

I wanted to chime into this thread with a few notes. On [Heroku's note](https://devcenter.heroku.com/articles/heroku-postgresql#multiple-schemas) about having 50+ schemas, another dimension that comes into the picture is the number of tables. For example, if you have 100 tables in each schema, your database would then need to support 5K tables (plus indexes if any). This may introduce performance issues in the following way:

- Client memory bloat: When you connect to Postgres, you open a separate process. Postgres caches data during the lifetime of a connection and the more objects you touch, the higher the memory requirements.
- Time it takes to do migrations: The more tables and indexes you have, the longer migrations take.
- Database backups: For each table, pg_dump walks over the entire set of tables and as a result backups may take a long time.

@bradrobertson has a pretty informative [blog post](https://influitive.io/our-multi-tenancy-journey-with-postgres-schemas-and-apartment-6ecda151a21f) on this topic and their experiences with scaling with this approach.

At Citus, we find that a good way to dynamically scale to thousands of tenants is by having all [tenants share the same tables](https://www.citusdata.com/blog/2016/10/03/designing-your-saas-database-for-high-scalability). This way, if you have customers who are on the free plan or sign up but don't return, you only pay the cost of inserting a few rows (rather than creating dozens of tables and indexes). Citus also comes with features such as [tenant isolation](https://www.citusdata.com/blog/2017/02/16/citus61-released/) and shard rebalancing to help with dynamic scaling.

If you're thinking of scaling a multi-tenant database, I'd love to catch up and exchange notes. In fact, we're planning to roll out a database plan specifically designed for this use case and I'd be happy to hear your take. Please feel free to drop me a note at ozgun at citusdata.com.

**Comment by vfonic** [comment]
@ozgune Thanks a lot for your reply! It really helped! I'll take a look at https://github.com/citusdata/activerecord-multi-tenant it looks like a potential solution. It also looks a bit young (don't get me wrong). Do you use it in production?

**Comment by mikecmpbll** [comment]
can't comment on pg specific issues such as the memory issue referenced by brad in the blog, as i'm using mysql. the other issues he mentioned though i don't really share. we were able to minimise the problem of migrations taking a while by using parallel migrations, and the backup time is a non-issue for us (we backup from a slave DB, it takes an hour or so for 4500 databases).

the tenant isolation that having 1 DB (or schema) per tenant affords us is priceless for our business, so the column scoping is a non-starter. as brad says in the blog; use UUIDs. there are also other huge advantages to having isolated tenants also, such as very simple backup restoration. consider having to restore an individual tenant in a foreign-key based tenanting model? i'm sure there are solutions for it but it's far more complicated.

i think it's worth emphasising that there's no problems that can't be addressed when using apartment to scale to many thousands of tenants, in my experience, and as we can testify in production.

as far as your 'free tenants never using it' problem; why not just have an eviction policy? seems only practical.

in summary:

- heroku restrictions; idk don't use it 💁
- migration time: parallel migrations
- backups: not an issue for us

we're working on the multi-server stuff, which'll give us a final piece in the scaling puzzle by allowing us to do application-level sharding. if anyone's reading this who has real world problems with scaling apartment, please comment or create issues, so we can pool together and find good solutions :).

**Comment by ozgune** [comment]
@vfonic -- Yup, we know several users who are using [activerecord-multi-tenant](https://github.com/citusdata/activerecord-multi-tenant) in production to scale to tens of thousands of tenants with Citus. I'm sure many others use it with single node Postgres. If you'd like, I could ping @lfittl to get a more detailed view on the gem's production users.

We also have a detailed presentation on [this topic](https://www.percona.com/live/17/sessions/designing-your-saas-database-scale-postgres). I'll upload it to the Postgres wiki and update my comment with the link. In the meantime, please feel free to hit us up if you're interested.

**Comment by mikecmpbll** [comment]
i'm going to close the issue, but feel free to continue the discussion.

**Comment by lfittl** [comment]
@vfonic We initially based activerecord-multi-tenant on the acts_as_tenant gem, which has been around for a while - both have a different approach to scale-out than apartment.

If you are interested in this topic further, there is a blog post that describes this approach in more detail: https://www.citusdata.com/blog/2017/01/05/easily-scale-out-multi-tenant-apps/

**Comment by bfcoder** [comment]
We have roughly 330 tenants. When we do a pg_dump we run out of shared memory. So I guess if you have a huge beefy machine you could have thousands of tenants. But that is so unnecessary, when a small postgres machine will suffice. :/

**Comment by mikecmpbll** [comment]
what's the reason that pg_dump uses so much memory, and how is it related to # of schemas?

**Comment by bfcoder** [comment]
Because, as far as I know, pg_dump/pg_dumpall has to create a lock for every sequence and table for every schema for every database. As I understand it, that can quickly run the machine out of memory with a huge database and a large number of schemas.
